### PR TITLE
chore: bring examples back into repository and add tests

### PR DESCRIPTION
**Pull request by iand**
Implements the approach discussed in #1052. Examples are now in a submodule with a replace directive that ensures they will be built using the version of go-libp2p in this repo.

Adds a test harness that enables some of the examples to be run as tests. The test harness reads log output and verifies expected log messages are present. Adding a test to an example is done by extracting the core of the example into a new function that can be called from main and from the harness. I have refactored a couple of examples to use helper functions to make them more amenable to external testing.

Fixes the relay example which was using relay discovery which was deprecated in 2020

Fixes libp2p/go-libp2p-examples#191

### CODE REVIEW

**Comment by iand** [comment]
To ease review, this diff shows just changes made to add tests to examples: https://github.com/iand/go-libp2p/compare/iand:integrate-examples...iand:test-examples

**Review by Stebalien — CHANGES_REQUESTED** [review]

**Comment by Stebalien** [comment]
Ok, tested with https://github.com/protocol/.github/pull/89.

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,13 @@
...
-examples/multicodecs/multicodecs
-.idea
\ No newline at end of file
+examples/http-proxy/http-proxy
+examples/libp2p-host/libp2p-host
+examples/multipro/multipro
+examples/pubsub/chat
+examples/relay/relay
+examples/routed-echo/routed-echo
+.idea
```
[the remark this answers is not in the document]
Comment (iand) [reply .gitignore:13]: I just followed existing style but will split into separate files

File: examples/.circleci/config.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+version: 2
```
[the remark this answers is not in the document]
Comment (iand) [reply examples/.circleci/config.yml:1]: done

File: examples/LICENSE [hunk 3]
```diff
@@ -0,0 +1,21 @@
+The MIT License (MIT)
```
[the remark this answers is not in the document]
Comment (iand) [reply examples/LICENSE:1]: done

File: README.md [hunk 4]
```diff
@@ -78,7 +78,7 @@ As new releases of go-libp2p are made available, you can upgrade your applicatio
 
 ### Examples
 
-Examples can be found in the [examples repo](https://github.com/libp2p/go-libp2p-examples).
+Examples can be found on the [examples folder](examples).
```
[the remark this answers is not in the document]
Comment (iand) [reply README.md:81]: done

File: README.md [hunk 5]
```diff
@@ -180,7 +180,6 @@ List of packages currently in existence for libp2p:
 | [`go-libp2p-http`](//github.com/libp2p/go-libp2p-http) | [![Travis CI](https://travis-ci.com/libp2p/go-libp2p-http.svg?branch=master)](https://travis-ci.com/libp2p/go-libp2p-http) | [![codecov](https://codecov.io/gh/libp2p/go-libp2p-http/branch/master/graph/badge.svg)](https://codecov.io/gh/libp2p/go-libp2p-http) | HTTP on top of libp2p streams |
 | **Testing and examples** |
 | [`go-libp2p-testing`](//github.com/libp2p/go-libp2p-testing) | [![Travis CI](https://travis-ci.com/libp2p/go-libp2p-testing.svg?branch=master)](https://travis-ci.com/libp2p/go-libp2p-testing) | [![codecov](https://codecov.io/gh/libp2p/go-libp2p-testing/branch/master/graph/badge.svg)](https://codecov.io/gh/libp2p/go-libp2p-testing) | a collection of testing utilities for libp2p |
-| [`go-libp2p-examples`](//github.com/libp2p/go-libp2p-examples) | [![Travis CI](https://travis-ci.com/libp2p/go-libp2p-examples.svg?branch=master)](https://travis-ci.com/libp2p/go-libp2p-examples) | [![codecov](https://codecov.io/gh/libp2p/go-libp2p-examples/branch/master/graph/badge.svg)](https://codecov.io/gh/libp2p/go-libp2p-examples) | go-libp2p examples and tutorials |
```
[the remark this answers is not in the document]
Comment (iand) [reply README.md:183]: done

**Review by marten-seemann — APPROVED** [review]
Let's merge this, even though the Unified CI Setup is not done yet. Will fix that soon.

**Pull request merged**